applications/serial_lte_modem/src/mqtt_c/slm_at_mqtt.c

```diff
@@ -362,111 +324,76 @@ static int client_init(void)
 		struct mqtt_sec_config *tls_config;
 
 		tls_config = &(client.transport).tls.config;
-		tls_config->peer_verify = TLS_PEER_VERIFY_REQUIRED;
-		tls_config->cipher_list = NULL;
-		tls_config->cipher_count = 0;
+		tls_config->peer_verify   = TLS_PEER_VERIFY_REQUIRED;
+		tls_config->cipher_list   = NULL;
+		tls_config->cipher_count  = 0;
 		tls_config->sec_tag_count = 1;
-		tls_config->sec_tag_list = (int *)&ctx.sec_tag;
-		tls_config->hostname = ctx.url;
-		client.transport.type = MQTT_TRANSPORT_SECURE;
+		tls_config->sec_tag_list  = (int *)&ctx.sec_tag;
+		tls_config->hostname      = mqtt_broker_url;
+		client.transport.type     = MQTT_TRANSPORT_SECURE;
 	} else {
-		client.transport.type = MQTT_TRANSPORT_NON_SECURE;
+		client.transport.type     = MQTT_TRANSPORT_NON_SECURE;
 	}
-
-	return err;
-}
-
-/**@brief Initialize the file descriptor structure used by poll.
- */
-static int fds_init(struct mqtt_client *c)
-{
-	if (c->transport.type == MQTT_TRANSPORT_NON_SECURE) {
-		fds.fd = c->transport.tcp.sock;
-	} else {
-#if defined(CONFIG_MQTT_LIB_TLS)
-		fds.fd = c->transport.tls.sock;
 #else
-		return -ENOTSUP;
+	client.transport.type = MQTT_TRANSPORT_NON_SECURE;
 #endif
-	}
-
-	fds.events = POLLIN;
-
-	return 0;
 }
 
 static int do_mqtt_connect(void)
 {
 	int err;
 
 	if (ctx.connected) {
-		return -EINPROGRESS;
+		return -EISCONN;
 	}
 
-	err = client_init();
-	if (err != 0) {
+	/* Init MQTT broker */
+	err = broker_init();
+	if (err) {
 		return err;
 	}
 
-#if defined(CONFIG_SLM_NATIVE_TLS)
-	if (ctx.sec_tag != INVALID_SEC_TAG) {
-		err = slm_tls_loadcrdl(ctx.sec_tag);
-		if (err < 0) {
-			LOG_ERR("Fail to load credential: %d", err);
-			return err;
-		}
-	}
-#endif
-
+	/* Connect to MQTT broker */
+	client_init();
 	err = mqtt_connect(&client);
 	if (err != 0) {
 		LOG_ERR("ERROR: mqtt_connect %d", err);
-#if defined(CONFIG_SLM_NATIVE_TLS)
-		if (ctx.sec_tag != INVALID_SEC_TAG) {
-			if (slm_tls_unloadcrdl(ctx.sec_tag) != 0) {
-				LOG_ERR("Fail to load credential: %d", err);
-			}
-		}
-#endif
 		return err;
 	}
 
-	err = fds_init(&client);
-	if (err != 0) {
-		LOG_ERR("ERROR: fds_init %d", err);
-		do_mqtt_disconnect();
-		return err;
-	}
+	k_thread_create(&mqtt_thread, mqtt_thread_stack,
+			K_THREAD_STACK_SIZEOF(mqtt_thread_stack),
+			mqtt_thread_fn, NULL, NULL, NULL,
+			THREAD_PRIORITY, K_USER, K_NO_WAIT);
 
-	/** start polling now for CONNACK */
 	ctx.connected = true;
-	k_sem_give(&mqtt_connected);
-
 	return 0;
 }
 
 static int do_mqtt_disconnect(void)
 {
 	int err;
 
+	if (!ctx.connected) {
+		return -ENOTCONN;
+	}
+
 	err = mqtt_disconnect(&client);
 	if (err) {
 		LOG_ERR("ERROR: mqtt_disconnect %d", err);
+		return err;
 	}
-#if defined(CONFIG_SLM_NATIVE_TLS)
-	if (ctx.sec_tag != INVALID_SEC_TAG) {
-		err = slm_tls_unloadcrdl(ctx.sec_tag);
-		if (err < 0) {
-			LOG_ERR("Fail to load credential: %d", err);
-			return err;
-		}
+
+	if (k_sem_take(&mqtt_disconnected, K_SECONDS(CONFIG_MQTT_KEEPALIVE)) != 0) {
```

Comment:
Not a strong push, since it works this way as well, but FYI there's `k_thread_join()` kernel function, which does exactly what is needed here (waits for some other thread to terminate).

Reply:
Thanks for the advice and I'll test.
This is a general scenario that a Kwork need to wait for the termination of a thread.